File: go/internal/e2e/rpc_session_state_extras_e2e_test.go
package e2e

import (
	"encoding/json"
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
	"github.com/github/copilot-sdk/go/rpc"
)

func TestRpcSessionStateExtras(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should_list_models_for_session", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const token = "rpc-session-model-list-token"
		registerProxyUser(t, ctx, token, "rpc-session-extras-user", nil)
		authClient := newAuthenticatedClient(ctx, token)
		defer authClient.ForceStop()

		session := createPortedSession(t, authClient, &copilot.SessionConfig{Model: "claude-sonnet-4.5"})
		defer session.Disconnect()

		result, err := session.RPC.Model.List(t.Context())
		if err != nil {
			t.Fatalf("Model.List failed: %v", err)
		}
		if result.List == nil {
			t.Fatal("Expected non-nil model list")
		}
		if len(result.List) == 0 {
			t.Fatal("Expected non-empty model list")
		}
		found := false
		for _, model := range result.List {
			data, err := json.Marshal(model)
			if err == nil && strings.Contains(string(data), "claude-sonnet-4.5") {
				found = true
				break
			}
		}
		if !found {
			t.Fatalf("Expected model list to include claude-sonnet-4.5, got %+v", result.List)
		}
	})

	t.Run("should_report_session_activity_when_idle", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		activity, err := session.RPC.Metadata.Activity(t.Context())
		if err != nil {
			t.Fatalf("Metadata.Activity failed: %v", err)
		}
		if activity.HasActiveWork {
			t.Fatal("Expected a fresh session to report no active work")
		}
		if activity.Abortable {
			t.Fatal("Expected a fresh session to have nothing abortable")
		}
	})

	t.Run("should_get_and_set_allowall_permissions", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()
		defer func() {
			_, _ = session.RPC.Permissions.SetAllowAll(t.Context(), &rpc.PermissionsSetAllowAllRequest{Enabled: false})
		}()

		initial, err := session.RPC.Permissions.GetAllowAll(t.Context())
		if err != nil {
			t.Fatalf("Permissions.GetAllowAll initial failed: %v", err)
		}
		if initial.Enabled {
			t.Fatal("Allow-all should be disabled on a fresh session")
		}

		enable, err := session.RPC.Permissions.SetAllowAll(t.Context(), &rpc.PermissionsSetAllowAllRequest{Enabled: true})
		if err != nil {
			t.Fatalf("Permissions.SetAllowAll(true) failed: %v", err)
		}
		if !enable.Success || !enable.Enabled {
			t.Fatalf("Expected successful enable, got %+v", enable)
		}
		afterEnable, err := session.RPC.Permissions.GetAllowAll(t.Context())
		if err != nil {
			t.Fatalf("Permissions.GetAllowAll after enable failed: %v", err)
		}
		if !afterEnable.Enabled {
			t.Fatal("Expected allow-all to be enabled")
		}

		disable, err := session.RPC.Permissions.SetAllowAll(t.Context(), &rpc.PermissionsSetAllowAllRequest{Enabled: false})
		if err != nil {
			t.Fatalf("Permissions.SetAllowAll(false) failed: %v", err)
		}
		if !disable.Success || disable.Enabled {
			t.Fatalf("Expected successful disable, got %+v", disable)
		}
		afterDisable, err := session.RPC.Permissions.GetAllowAll(t.Context())
		if err != nil {
			t.Fatalf("Permissions.GetAllowAll after disable failed: %v", err)
		}
		if afterDisable.Enabled {
			t.Fatal("Expected allow-all to be disabled")
		}
	})

	t.Run("should_read_empty_sql_todos_for_fresh_session", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		result, err := session.RPC.Plan.ReadSqlTodos(t.Context())
		if err != nil {
			t.Fatalf("Plan.ReadSqlTodos failed: %v", err)
		}
		if result.Rows == nil {
			t.Fatal("Expected non-nil SQL todo rows")
		}
		if len(result.Rows) != 0 {
			t.Fatalf("Expected empty SQL todo rows, got %+v", result.Rows)
		}
	})

	t.Run("should_get_telemetry_engagement_id", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		result, err := session.RPC.Telemetry.GetEngagementId(t.Context())
		if err != nil {
			t.Fatalf("Telemetry.GetEngagementId failed: %v", err)
		}
		if result == nil {
			t.Fatal("Expected non-nil telemetry engagement result")
		}
	})

	t.Run("should_get_current_tool_metadata_after_initialization", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		answer, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 2+2?"})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}
		if answer == nil {
			t.Fatal("Expected a final assistant message")
		}

		result, err := session.RPC.Tools.GetCurrentMetadata(t.Context())
		if err != nil {
			t.Fatalf("Tools.GetCurrentMetadata failed: %v", err)
		}
		if result.Tools == nil {
			t.Fatal("Expected non-nil current tool metadata")
		}
		if len(result.Tools) == 0 {
			t.Fatal("Expected non-empty current tool metadata")
		}
		for _, tool := range result.Tools {
			if strings.TrimSpace(tool.Name) == "" {
				t.Fatalf("Expected non-empty tool name, got %+v", tool)
			}
			if strings.TrimSpace(tool.Description) == "" {
				t.Fatalf("Expected non-empty tool description, got %+v", tool)
			}
		}
	})

	t.Run("should_reload_session_plugins", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		if _, err := session.RPC.Plugins.Reload(t.Context()); err != nil {
			t.Fatalf("Plugins.Reload failed: %v", err)
		}
		plugins, err := session.RPC.Plugins.List(t.Context())
		if err != nil {
			t.Fatalf("Plugins.List failed: %v", err)
		}
		if plugins.Plugins == nil {
			t.Fatal("Expected non-nil session plugin list")
		}
		for _, plugin := range plugins.Plugins {
			if strings.TrimSpace(plugin.Name) == "" {
				t.Fatalf("Expected non-empty plugin name, got %+v", plugin)
			}
		}
	})
}

File: nodejs/test/e2e/rpc_session_state_extras.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import type { CopilotSession } from "../../src/index.js";
import { approveAll, CopilotClient, RuntimeConnection } from "../../src/index.js";
import { createSdkTestContext, DEFAULT_GITHUB_TOKEN } from "./harness/sdkTestContext.js";

describe("Session-scoped state extras RPC", async () => {
    const { copilotClient: client, env, openAiEndpoint, workDir } = await createSdkTestContext();

    function createClientWithEnv(
        extraEnv: Record<string, string | undefined>,
        token = DEFAULT_GITHUB_TOKEN
    ): CopilotClient {
        return new CopilotClient({
            workingDirectory: workDir,
            env: {
                ...env,
                ...extraEnv,
            },
            logLevel: "error",
            connection: RuntimeConnection.forStdio({ path: process.env.COPILOT_CLI_PATH }),
            gitHubToken: token,
        });
    }

    function createAuthenticatedClient(token: string): CopilotClient {
        return createClientWithEnv(
            {
                COPILOT_DEBUG_GITHUB_API_URL: env.COPILOT_API_URL,
            },
            token
        );
    }

    async function configureAuthenticatedUser(token: string): Promise<void> {
        await openAiEndpoint.setCopilotUserByToken(token, {
            login: "rpc-session-extras-user",
            copilot_plan: "individual_pro",
            endpoints: {
                api: env.COPILOT_API_URL,
                telemetry: "https://localhost:1/telemetry",
            },
            analytics_tracking_id: "rpc-session-extras-tracking-id",
        });
    }

    async function createSession(): Promise<CopilotSession> {
        return client.createSession({ onPermissionRequest: approveAll });
    }

    async function disconnect(session: CopilotSession | undefined): Promise<void> {
        if (!session) {
            return;
        }
        try {
            await session.disconnect();
        } catch {
            // Best-effort cleanup.
        }
    }

    it("should list models for session", { timeout: 120_000 }, async () => {
        const token = "rpc-session-model-list-token";
        await configureAuthenticatedUser(token);
        const authClient = createAuthenticatedClient(token);
        let session: CopilotSession | undefined;
        try {
            await authClient.start();
            session = await authClient.createSession({
                model: "claude-sonnet-4.5",
                onPermissionRequest: approveAll,
            });

            const result = await session.rpc.model.list();

            expect(Array.isArray(result.list)).toBe(true);
            expect(result.list.length).toBeGreaterThan(0);
            expect(
                result.list.some((model) => JSON.stringify(model).includes("claude-sonnet-4.5"))
            ).toBe(true);
        } finally {
            await disconnect(session);
            try {
                await authClient.forceStop();
            } catch {
                // Best-effort cleanup.
            }
        }
    });

    it("should report session activity when idle", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const activity = await session.rpc.metadata.activity();

            expect(activity.hasActiveWork).toBe(false);
            expect(activity.abortable).toBe(false);
        } finally {
            await session.disconnect();
        }
    });

    it("should get and set allowall permissions", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const initial = await session.rpc.permissions.getAllowAll();
            expect(initial.enabled).toBe(false);

            const enable = await session.rpc.permissions.setAllowAll({ enabled: true });
            expect(enable.success).toBe(true);
            expect(enable.enabled).toBe(true);
            expect((await session.rpc.permissions.getAllowAll()).enabled).toBe(true);

            const disable = await session.rpc.permissions.setAllowAll({ enabled: false });
            expect(disable.success).toBe(true);
            expect(disable.enabled).toBe(false);
            expect((await session.rpc.permissions.getAllowAll()).enabled).toBe(false);
        } finally {
            try {
                await session.rpc.permissions.setAllowAll({ enabled: false });
            } catch {
                // Best-effort reset.
            }
            await session.disconnect();
        }
    });

    it("should read empty sql todos for fresh session", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const result = await session.rpc.plan.readSqlTodos();

            expect(result.rows).toBeDefined();
            expect(result.rows).toEqual([]);
        } finally {
            await session.disconnect();
        }
    });

    it("should get telemetry engagement id", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const result = await session.rpc.telemetry.getEngagementId();

            expect(result).toBeDefined();
        } finally {
            await session.disconnect();
        }
    });

    it("should get current tool metadata after initialization", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const answer = await session.sendAndWait({ prompt: "What is 2+2?" });
            expect(answer).toBeDefined();

            const result = await session.rpc.tools.getCurrentMetadata();

            expect(result.tools).not.toBeNull();
            expect(result.tools!.length).toBeGreaterThan(0);
            for (const tool of result.tools!) {
                expect(tool.name).toBeTruthy();
                expect(tool.description).toBeDefined();
            }
        } finally {
            await session.disconnect();
        }
    });

    it("should reload session plugins", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            await session.rpc.plugins.reload();

            const plugins = await session.rpc.plugins.list();
            expect(plugins.plugins).toBeDefined();
            for (const plugin of plugins.plugins) {
                expect(plugin.name).toBeTruthy();
            }
        } finally {
            await session.disconnect();
        }
    });
});
